### PR TITLE
add ceph csi driver proposal

### PR DESCRIPTION
**Pull request by rootfs**
**Description of your changes:**

Ceph CSI driver support rook

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**
- [ ] Documentation has been updated, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.
- [ ] `make vendor` does not cause changes.
- [ ] Comments have been added or updated based on the standards set in [CONTRIBUTING.md](../blob/master/CONTRIBUTING.md#comments)

@travisn @BlaineEXE @ganeshmaharaj @gman0 @mkimuram

[skip ci]

### CODE REVIEW

**Comment by galexrt** [comment]
@rootfs Please add `[skip ci]` to the PR description when it is just a doc change/design document. I have added it and re-run CI (CI is green now).

**Comment by travisn** [comment]
@rootfs Could you expand on the details of what it will look like for Rook to integrate the csi driver? Here are a few questions:
- Will the ceph-csi driver work both for Rook and an external Ceph cluster? It's not clear to me why that will work since in the past the in-tree drivers had some differences from the Rook flex driver.
- For some time we will need to support both the flex driver and a csi driver. These will be independent to support, right?
- What will the flow look like in Rook to setup the CSI driver? Will the operator register it?
- Do you have a prototype of Rook working with the csi driver? That would help show the design.

**Comment by bassam** [comment]
I'm also interested in understanding if the rook-agent will be replaced by ceph-csi completely or still have a role. We've envisioned the rook-agent as handling a number of tasks on the consumption side of storage that might not be in scope for CSI. Some examples:

- launching user-mode proxies (like nbd-runner and tcmu etc.)
- running sidecars just as with the Cassandra proposal

Essentially the rook-agent acts as a client side mini-controller that facilitates/improves the overall experience. If we keep developing this role of the rook-agent, it would seem to me that the ceph-csi is only part of the story and not a replacement for rook-agent.

File: design/ceph-csi-driver.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+Both Ceph RBD and Ceph FS drivers can be found at [ceph/ceph-csi](https://github.com/ceph/ceph-csi). Currently the drivers are up to CSI v0.3.0 spec.
+
+* RBD driver. Currently, rbd driver supports krbd. There is a consideration to support other forms of drivers such as rbd-nbd or TCMU based drivers.
+* CephFS driver. Both Kernel Ceph FS and Ceph FUSE are supported. When `ceph-fuse` is installed, it is chosen as default mounter.
+
+There is also upstream Kubernetes work to [include these drivers for e2e tests](https://github.com/kubernetes/kubernetes/pull/67088).
+
+## Kubernetes CSI Driver Deployment
+
+Staring Kubernetes CSI driver brings up an [external-provisioner](https://github.com/kubernetes-csi/external-provisioner), an [external-attacher](https://github.com/kubernetes-csi/external-attacher), DaemonSet that runs the driver on the nodes, and optionally an [external-snapshotter](https://github.com/kubernetes-csi/external-snapshotter).
```
Comment (galexrt) [inline design/ceph-csi-driver.md]: `Staring` -> `Starting`

File: design/ceph-csi-driver.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+common storage features such as create/delete volumes, publish/unpublish volumes, stage/unstage volumes, and more. It is currently at 0.3.0 release.
+
+Kubernetes started to CSI driver alpha support in [1.9](https://kubernetes.io/blog/2018/01/introducing-container-storage-interface/), beta support in [1.10](https://kubernetes.io/blog/2018/04/10/container-storage-interface-beta/).
+
+It is projected that CSI will be the only supported persistent storage driver
+in the near feature. In-tree drivers such as Ceph RBD and Ceph FS will be replaced with their respective CSI drivers.
+
+## Ceph CSI Drivers Status
+
+There have been active Ceph CSI drivers development since Kubernetes 1.9.
```
Comment (galexrt) [inline design/ceph-csi-driver.md]: `have` -> `has`

File: design/ceph-csi-driver.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+Kubernetes started to CSI driver alpha support in [1.9](https://kubernetes.io/blog/2018/01/introducing-container-storage-interface/), beta support in [1.10](https://kubernetes.io/blog/2018/04/10/container-storage-interface-beta/).
+
+It is projected that CSI will be the only supported persistent storage driver
+in the near feature. In-tree drivers such as Ceph RBD and Ceph FS will be replaced with their respective CSI drivers.
+
+## Ceph CSI Drivers Status
+
+There have been active Ceph CSI drivers development since Kubernetes 1.9. 
+Both Ceph RBD and Ceph FS drivers can be found at [ceph/ceph-csi](https://github.com/ceph/ceph-csi). Currently the drivers are up to CSI v0.3.0 spec.
```
Comment (galexrt) [inline design/ceph-csi-driver.md]: Isn't CephFS always written as one word `CephFS`?

> Chapter 1. What is the Ceph File System (CephFS)? - Red Hat[...]

File: design/ceph-csi-driver.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+For example, deploying a CephFS CSI driver consists of the following steps:
+1. Creating a [RBAC for external provisioner](https://github.com/ceph/ceph-csi/blob/master/deploy/cephfs/kubernetes/csi-provisioner-rbac.yaml) and the [provisioner itself](https://github.com/ceph/ceph-csi/blob/master/deploy/cephfs/kubernetes/csi-cephfsplugin-provisioner.yaml).
+2. Creating a [RBAC for external attacher](https://github.com/ceph/ceph-csi/blob/master/deploy/cephfs/kubernetes/csi-attacher-rbac.yaml) and [attacher itself](https://github.com/ceph/ceph-csi/blob/master/deploy/cephfs/kubernetes/csi-attacher-rbac.yaml).
+3. Creating [RBAC for CSI driver](https://github.com/ceph/ceph-csi/blob/master/deploy/cephfs/kubernetes/csi-nodeplugin-rbac.yaml) and [driver DaemonSet](https://github.com/ceph/ceph-csi/blob/master/deploy/cephfs/kubernetes/csi-cephfsplugin.yaml)
+
+## Rook Integration
+
+Rook's Local Ceph Node Agent is modeled after flexvolume drivers. There are some overlapping with Kubernetes CSI model:
```
Comment (galexrt) [inline design/ceph-csi-driver.md]: `Local` -> `local`

File: design/ceph-csi-driver.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+- The flexvolume drivers can be replaced by CSI drivers when running on Kubernetes 1.10+.
+
+## Integration Plan
+
+Rook Ceph flexdriver supports `rbd-nbd`, but ceph-csi rbd driver doesn't. This gap needs to be scoped in ceph-csi driver.
+
+Once it deploys the CSI driver and external controller, rook operator creates a new storage class that uses CSI provisioners,
+which in turn provisions PVs through CSI drivers. This will not overlap with existing flex driver based PVs.
+
+
```
Comment (galexrt) [inline design/ceph-csi-driver.md]: Could you remove the second empty line.

File: design/ceph-csi-driver.md [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+- VolumeAttachment API objects are currently in Kubernetes storage API group. The VolumeAttachment watchers and finalizers can be replaced with CSI attacher.
+- There is external CSI volume provisioner that calls out the drivers' create/delete volume. Rook's own provisioner needs to be integrated/replaced with the external provisioner
+- The flexvolume drivers can be replaced by CSI drivers when running on Kubernetes 1.10+.
+
+## Integration Plan
+
+Rook Ceph flexdriver supports `rbd-nbd`, but ceph-csi rbd driver doesn't. This gap needs to be scoped in ceph-csi driver.
+
+Once it deploys the CSI driver and external controller, rook operator creates a new storage class that uses CSI provisioners,
```
Comment (galexrt) [inline design/ceph-csi-driver.md]: `rook` -> `Rook`

**Comment by rootfs** [comment]
@galexrt thanks for the review
cc @travisn @ganeshmaharaj @gman0

File: design/ceph-csi-driver.md [hunk 7]
```diff
@@ -8,21 +8,21 @@ common storage features such as create/delete volumes, publish/unpublish volumes
...
 ## Ceph CSI Drivers Status
 
-There have been active Ceph CSI drivers development since Kubernetes 1.9. 
-Both Ceph RBD and Ceph FS drivers can be found at [ceph/ceph-csi](https://github.com/ceph/ceph-csi). Currently the drivers are up to CSI v0.3.0 spec.
+There have been active Ceph CSI drivers developments since Kubernetes 1.9. 
+Both Ceph RBD and CephFS drivers can be found at [ceph/ceph-csi](https://github.com/ceph/ceph-csi). Currently the drivers are up to CSI v0.3.0 spec.
 
 * RBD driver. Currently, rbd driver supports krbd. There is a consideration to support other forms of drivers such as rbd-nbd or TCMU based drivers.
-* CephFS driver. Both Kernel Ceph FS and Ceph FUSE are supported. When `ceph-fuse` is installed, it is chosen as default mounter.
+* CephFS driver. Both Kernel CephFS and Ceph FUSE are supported. When `ceph-fuse` is installed, it is chosen as default mounter.
```
Comment (galexrt) [inline design/ceph-csi-driver.md]: Where must `ceph-fuse` be installed? Inside the Ceph CSI container, the host or kubelet (container)?

Comment (kfox1111) [reply design/ceph-csi-driver.md]: in the csi driver container.

File: design/ceph-csi-driver.md [hunk 8]
```diff
@@ -0,0 +1,47 @@
+# Ceph CSI Driver Support
+
```
Comment (travisn) [inline design/ceph-csi-driver.md]: Could you add a comment about what rook release we would be targeting? (similar to "Targeted for v0.9" as seen [here](https://github.com/rook/rook/blob/master/design/decouple-ceph-version.md)). Even though we are dependent on CSI stabilizing, could we target 0.9 to support both CSI and Flex?

**Pull request merged**